[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115877
* #115814
* #115813
* #115655
* #115654
* #112861
* #114687
* #115270
* #115174
* #115072
* #114606
* #114319
* #114075
* #113932
* #113891
* #113959
* #113890
* #113570
* #113569
* #113558
* #113399
* __->__ #113366
* #113365

When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

kadeng committed:
[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -419,6 +419,53 @@ def swap_XW(
+        if b_layout is not None:
+            new_op.B.layout = CUTLASSGemmTemplate.cutlass_layout(b_layout)
+        if c_layout is not None:
+            new_op.C.layout = CUTLASSGemmTemplate.cutlass_layout(d_layout)
```
Should be `c_layout` here?

aakhundov commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -660,7 +707,10 @@ def render(  # type: ignore[override]
-
+        # The layouts might have changed between autotuning and this call if they were FlexibleLayout
+        # we need to adapt, which might lead to suboptimal performance.
+        # @TODO kadeng: Find a way to solve this better
```
I believe, this `TODO` is sufficient :)

kadeng replied:
Just FYI: This is addressed later in the diff stack..

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -419,6 +437,51 @@ def swap_XW(
+        all_match = all(
+            CUTLASSGemmTemplate.layout_match(buf.get_layout(), op_layout)
+            for buf, op_layout in zip(
+                [X, W, Bias, Y],
+                [op.A.layout, op.B.layout, op.C.layout, op.D.layout],
```
```suggestion
                (X, W, Bias, Y),
                (op.A.layout, op.B.layout, op.C.layout, op.D.layout),
```

kadeng replied:
done

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render"


When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented:
Moved to a (draft) feature branch, see https://github.com/pytorch/pytorch/pull/115919
